packages/flame/lib/src/experimental/layout_component.dart

```diff
@@ -0,0 +1,123 @@
+import 'package:flame/components.dart';
+import 'package:flame/experimental.dart';
+import 'package:flutter/widgets.dart';
+
+enum Direction { horizontal, vertical }
+
+/// Super class for [RowComponent] and [ColumnComponent]
+/// A relayout is performed when
+///  - a change in this component's children takes place
+///  - the [gap] parameter is changed
+abstract class LayoutComponent extends PositionComponent with HasGameRef {
+  LayoutComponent(this.direction, this.mainAxisAlignment, this._gap);
+  final Direction direction;
+  final MainAxisAlignment mainAxisAlignment;
+
+  /// gap between components
+  double _gap;
+
+  set gap(double gap) {
+    _gap = gap;
+    layoutChildren();
+  }
+
+  double get gap => _gap;
+
+  @override
+  void onChildrenChanged(Component child, ChildrenChangeType changeType) {
+    layoutChildren();
+  }
+
+  void layoutChildren() {
+    final list = children.whereType<PositionComponent>().toList();
+    if (list.isEmpty) {
+      return;
+    }
+    final currentPosition = Vector2.zero();
+    final componentsDimension = list.fold<double>(
```

Comment:
This should be set to the `PositionComponent`'s width or height right? And then the one that isn't set here should be the `max` width/height of the child components.

Reply:
The size of the parent component is checked and handled here:
https://github.com/flame-engine/flame/blob/7486750248cba830bfb900632f9471fa9635d375/packages/flame/lib/src/experimental/layout_component.dart#L45-L48

Reply:
`dimensionAvailable` should possibly be the parent's `size`, or it's own `size`, it shouldn't check anything regarding the canvas size or such.
So either we let the `size` of the `LayoutComponent` grow dynamically with the children added to it, or it has a set size from the beginning, but to make it more flutter like I think that it should set the size itself and then just be constrained by the parent's `size`.
Because right now the layout component's `size` will be (0, 0) now since it isn't updated when children are added. The setter of `size` should probably also be overridden so that it can't be set too if we choose to go with a dynamic size.

Reply:
But can the parent not also have a zero size?
I had it in mind like this: if the developer defines a size for LayoutComponent then this size will be taken for the available dimension calculation, else it just takes the canvas' width or height for the end edge.

Reply:
Why would it take the canvas size? A component can live anywhere in the tree.
To lower the confusion as much as possible for these ones I think we should follow the Flutter style with having the constraints go down and the sizes go up (or in this case just set it itself).

Reply:
Ok, but how would you retrieve the width (or height) for the case if the developer hasn't defined an explicit size for the `LayoutComponent` in order for being able to have a value for `dimensionAvailable`?
`parent.size` cannot be used because `parent` is a `Component` and doesn't necessarily have a `size`.

Reply:
... and growing with the size of its children doesn't make much sense for alignment types like spaceBetween imo.

Reply:
You take the first available parent that is a `PositionComponent` (or a `SizeProvider` when we have that, not sure why we don't have that currently) with `findParent<PositionComponent>` and if that returns null then the size is unbounded, so the row or column can grow infinitely.

Reply:
Meanwhile i've changed it to this:
```
/// Either we have a defined size or we just use the summed width/height of
/// the current children.
final dimensionAvailable = size[vectorIndex] != 0.0
    ? size[vectorIndex]
    : componentsDimension;
```
With your suggestion you wouldn't allow to set an explicit `size` for the `LayoutComponent`, but only for its parents up the hierarchy.
What if i would check for an explicit `size` of the `LayoutComponent` _and_ if that is zero look for a `size` in the first found `SizeProvider` up the hierarchy? I wouldn't forbid setting a `size` for the `LayoutComponent`.

Reply:
> What if i would check for an explicit size of the LayoutComponent and if that is zero look for a size in the first found SizeProvider up the hierarchy? I wouldn't forbid setting a size for the LayoutComponent.

I think this will be prone to bugs, what if the user shrinks the size down to 0? Then it will all of a sudden start using `componentsDimension` instead. Maybe take in a boolean instead that specifies whether it should be manually or automatically sized? That boolean could be set to true if a `size` is sent in through the constructor.

We'd also have to connect a listener to `size` in the cases that it is manually sized, and in the cases that it is automatically sized we have to connect a listener to the ancestor size, so that we can do `layoutChildren` when the bounds change.

Reply:
It gets more and more complicated...
Pasha suggested that the developer should call layoutChildren if the size is manually changed, i.e. not to use a listener for that case.

Reply:
I don't see the point of laying that responsibility on the user when we have the possibility to easily add a listener to the bounds with:
```dart
final bounds = isManuallySized ? size : findParent<PositionComponent>;
bounds.addListener(layoutChildren);
```

Reply:
Perhaps let's not overcomplicate this?

In the first version of the LayoutComponent the `size` will have to be provided explicitly. In the next version (in a separate PR), we will relax that requirement and allow `size` to be omitted, in which case it will be derived from a parent.

Reply:
As you like. It's already implemented though.

Reply:
> Perhaps let's not overcomplicate this?
> 
> In the first version of the LayoutComponent the `size` will have to be provided explicitly. In the next version (in a separate PR), we will relax that requirement and allow `size` to be omitted, in which case it will be derived from a parent.

When re-reviewing this I agree, maybe it's a bit overkill to listen to the parent's size.
Then we can scrap `isManuallySized` too, so that the component only grows to its size, can we restrict the user from changing the `size` variable from the outside though? Then we only need to do the layout when children are added.

Later we could add a `WrapComponent` so that components can be organized within a set size.

Reply:
What do you think about this @rivella50? :)

Reply:
@spydon Sure, why not. I've already removed setting the `size` and `isManuallySized` last weekend.
Just to be clear: Without being able to set the `size` (parent or `LayoutComponent`) the alignment types spaceXY don't make sense anymore.

Reply:
> Without being able to set the size (parent or LayoutComponent) the alignment types spaceXY don't make sense anymore.

Hmm, that's true, a shame to drop that one...
But maybe it is worth dropping it to just have a simple first version?